manual/manual/cmds/intf-c.etex

```diff
@@ -483,15 +483,15 @@ cast to and from the type "value". This includes pointers returned by
 obtained with the \verb'&' operator.
    %%% FIXME will change in 4.02.0 (?)
 
-Caution: if a pointer returned by "malloc" is cast to the type "value"
-and returned to OCaml, explicit deallocation of the pointer using
-"free" is potentially dangerous, because the pointer may still be
-accessible from the OCaml world.  Worse, the memory space deallocated
-by "free" can later be reallocated as part of the OCaml heap; the
-pointer, formerly pointing outside the OCaml heap, now points inside
-the OCaml heap, and this can crash the garbage collector.  To avoid
-these problems, it is preferable to wrap the pointer in a OCaml block
-with tag "Abstract_tag" or "Custom_tag".
+Caution: if a pointer returned by "malloc" is cast to the type "value" and
+returned to OCaml, explicit deallocation of the pointer using "free" is
+potentially dangerous, because the pointer may still be accessible from the
+OCaml world.  Worse, the memory space deallocated by "free" can later be
+reallocated as part of the OCaml heap; the pointer, formerly pointing outside
+the OCaml heap, now points inside the OCaml heap, and this can crash the
+garbage collector.  To avoid these problems, it is recommended to wrap the
+pointer in an OCaml block with header "Caml_out_of_heap_header" defined in
+"caml/mlvalues.h".
```

Comment:
With this documentation change, the suggestion to use Abstract_tag or Custom_tag is gone. Is it not the recommended approach anymore (is it only correct in presence of the page table?)? Is Caml_out_of_heap_header  a generalization, or complementary?

My non-expert understanding is that Abstract_tag and Custom_tag are no-scan-tags, so the GC will not traverse the block (whatever the liveness property of the block is), and Caml_out_of_heap_header can contain arbitrary tags but are always marked already-traversed for the GC.

Reply:
Note: I think that the example struct you had in the PR description may be useful in the documentation as well.

Reply:
With naked pointers, the only requirement is that the headers are marked black. It doesn't matter what the tag is. The issue with `Abstract_tag` and `Custom_tag` with the color being non-black is that the GC will write to the header while marking. This may not be appropriate if the out-of-heap region is marked as a read-only section. This was pointed out by @stedolan.

I don't want to recommend `Custom_tag` since a custom object is expected to have a block of custom operations as its first field. Not having the correct first field may cause a segfault on marshall and polymorphic comparison. Better to recommend `Abstract_tag` as the code will get an `Invalid_argument` exception on marshall and polymorphic comparison. 

I'll edit the documentation appropriately now.

Reply:
So if I understand correctly, we need to use `Caml_out_of_heap_header` to avoid GC-related traversals, but we also need to pick the right tag for the other runtime operations (comparison/hashing), `Abstract_tag` being recommended as the default blackbox.

(Is there an approved/recommended way to masquerade pointers as non-pointers by tagging them? We use this technique in a couple places in the runtime, and I suppose it would be useful (and probably already used) for some FFI situations. When would one use this rather than a Black Abstract?)

Reply:
> So if I understand correctly, we need to use Caml_out_of_heap_header to avoid GC-related traversals, but we also need to pick the right tag for the other runtime operations (comparison/hashing), Abstract_tag being recommended as the default blackbox.

Yes. For non-expert users, who may want to just have a pointer to an out-of-heap structure, `Abstract_tag` is appropriate. Operations such as comparison and hashing will raise an exception. If you do want hashing and comparison to work, then (the expert) users will need to choose the right tag and size and corresponding object layout based on the need.

> (Is there an approved/recommended way to masquerade pointers as non-pointers by tagging them? We use this technique in a couple places in the runtime, and I suppose it would be useful (and probably already used) for some FFI situations. When would one use this rather than a Black Abstract?)

We use this in multicore bytecode runtime to tag the code pointers so that they're not followed by the GC (multicore does not use page tables). https://github.com/ocaml-multicore/ocaml-multicore/blob/172fe18bbc7fcdaa1fc286869c28a2c59ecbf27a/runtime/caml/mlvalues.h#L265-L281. This is an example of a situation where tagging rather than having a Black Abstract header makes sense. That said, we should discuss pointer tagging in a separate issue.